internal/server/cert.go

```diff
@@ -0,0 +1,37 @@
+package server
+
+import (
+	"crypto/tls"
+	"log/slog"
+)
+
+type CertManager interface {
+	GetCertificate(hello *tls.ClientHelloInfo) (*tls.Certificate, error)
+}
+
+type StaticCertManager struct {
+	tlsCertificateFilePath string
+	tlsPrivateKeyFilePath  string
+}
+
+func NewStaticCertManager(tlsCertificateFilePath, tlsPrivateKeyFilePath string) *StaticCertManager {
+	return &StaticCertManager{
+		tlsCertificateFilePath: tlsCertificateFilePath,
+		tlsPrivateKeyFilePath:  tlsPrivateKeyFilePath,
+	}
+}
+
+func (m *StaticCertManager) GetCertificate(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+	slog.Info(
+		"Loading custom TLS certificate",
+		"tls-certificate-path", m.tlsCertificateFilePath,
+		"tls-private-key-path", m.tlsPrivateKeyFilePath,
+	)
+
+	cert, err := tls.LoadX509KeyPair(m.tlsCertificateFilePath, m.tlsPrivateKeyFilePath)
```

Comment:
This will be called for every TLS handshake, which means the certificate files will be loaded multiple times. Better to load the certificate when the `StaticCertManager` is established, and reuse it for the lifetime of the deployment.

That would also let us catch a loading error early on, and fail the deployment at that point, which would be safer.

Reply:
Applied the patch and added a test. Let me know what you think

Reply:
Thinking about this, might need a mutex since there might be multiple procs loading certs at the same time.

Reply:
Confirmed, will work on a fix:

```go
	manager := NewStaticCertManager(certPath, keyPath)
	go func() {
		manager.GetCertificate(&tls.ClientHelloInfo{})
	}()
	cert, err := manager.GetCertificate(&tls.ClientHelloInfo{})

```

Output of `go test ./... -race`:

```
WARNING: DATA RACE
Write at 0x00c0001e8e00 by goroutine 23:
  github.com/basecamp/kamal-proxy/internal/server.(*StaticCertManager).GetCertificate()
      /Users/dmytro/work/github/kamal-proxy/internal/server/cert.go:41 +0x254
  github.com/basecamp/kamal-proxy/internal/server.TestCertificateLoadingRaceCondition.func1()
      /Users/dmytro/work/github/kamal-proxy/internal/server/cert_test.go:49 +0x7c

Previous write at 0x00c0001e8e00 by goroutine 22:
  github.com/basecamp/kamal-proxy/internal/server.(*StaticCertManager).GetCertificate()
      /Users/dmytro/work/github/kamal-proxy/internal/server/cert.go:41 +0x254
  github.com/basecamp/kamal-proxy/internal/server.TestCertificateLoadingRaceCondition()
      /Users/dmytro/work/github/kamal-proxy/internal/server/cert_test.go:51 +0x270
  testing.tRunner()
      /Users/dmytro/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.1.darwin-arm64/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /Users/dmytro/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.23.1.darwin-arm64/src/testing/testing.go:1743 +0x40
```

Reply:
Introduced an RWMutex in https://github.com/basecamp/kamal-proxy/pull/17/commits/f1df35f7efc980e64a750e2c78a6e2933b628562 to address it

Reply:
Another option to avoid the mutex would be to attempt the file load in the `NewStaticCertManager`, and then save both the error and the response in the struct. `GetCertificate` would just return both, avoiding a delayed load, the need for mutex, etc.

This design might limit the options for (potential) future changes, like reloading the certificate on file change.